ironforge-cloud/geyser-kafka
# feat: emit account removal events

## Summary

Sending an account update with `lamports: 0` for accounts that we want deleted.

## Changes to Config

Add
```json
{
  "publish_account_deletions": true
}
```

to the global config. We don't want to turn this on by default since it requires us to listen
to `notify_transaction` to detect account deletions. See [this local config as an example](https://github.com/ironforge-cloud/geyser-kafka/blob/ffbf8cedecee357caf98436ce1eac4a910ea8820/config-local.json)

See _Details_.

## Details

Account delettions don't cause an account update to be emitted, therefore we get it from
transactions instead.

### Account Update from Transaction

- when we find a transaction where an account's lamports went to 0 we declare it as _deleted_
- we then assume that all other accounts could be its owner
- for each owner that a publisher is interested in we publish an update with that account with
  its lamports set to `0`
- tools listening to these events then know that this account was deleted
- deriving the `write_version` by incrementing the last one that we updated an account update
  with (since transactions don't have a `write_version` but we still need to know the order of
  operations in case an account is deleted and re-created in the same slot, or the other way
  around)

### Testing

This behavior is difficult to verify with unit tests therefore I added the following to allow
integration tests on the devs machine:

- instructions on how to setup sibling projects to help running tests and verifying results
- make tasks to perform one task at a time in separate terminals
- a full integration test which does the following:
  1. run the geyser-store binary which will listen for account + transaction updates via http
  and is able to return updates that it stored in a sqlite database
  2. startup amman with the geyser plugin active (using a config that points to the
  geyser-store)
  3. run the solanax scripts that create and delete a post
  4. shut down all tools started up

NOTE: the amman which supports geyser plugin configs is not yet released thus this works
only for my own testing ATM and/or for anyone who can install the newer amman version.

### Improvements not directly Related

#### Improved wanted transaction detection

- before if any account was not wanted by a publisher it would be
  ignored
- now if none of the accounts are not wanted by a publisher, the
  transaction is ignored

#### Logging

Improved logging of ignored accounts updates and transactions to make diagnostics easier.
